src/processor/operator/aggregate/aggregate_hash_table.cpp

```diff
@@ -579,48 +577,77 @@ void AggregateHashTable::findHashSlots(const FactorizedTable& srcTable, uint64_t
     }
 }
 
-void AggregateHashTable::updateAggState(const std::vector<ValueVector*>& flatKeyVectors,
-    const std::vector<ValueVector*>& unFlatKeyVectors, AggregateFunction& aggregateFunction,
-    ValueVector* aggVector, uint64_t multiplicity, uint32_t /*colIdx*/, uint32_t aggStateOffset) {
+void AggregateHashTable::appendDistinct(const std::vector<ValueVector*>& keyVectors,
+    ValueVector* aggregateVector) {
+    std::vector<ValueVector*> distinctKeyVectors(keyVectors);
+    distinctKeyVectors.push_back(aggregateVector);
+    if (!keyVectors.empty() && !aggregateVector->state->isFlat() &&
+        getFirstUnflatState(keyVectors) != aggregateVector->state.get()) {
+        // If the aggregateVector is unFlat, with a different state from the unflat vectors in the
+        // keys, flatten the aggregate vector and append values one by one since there can be at
+        // most one unflat group in the keyVectors.
+        auto oldState = std::move(aggregateVector->state);
```

Comment:
Instead of doing the crossproduct there, i suggest we flat the aggreagate vectors when aggregateVector and keyVector are both unflat and from different chunks.

Reply:
I'm also curious if we do flatten all but one, how much performance difference did you observe?

Reply:
I didn't know how to do that, but I think I've figured it out.

Reply:
> I'm also curious if we do flatten all but one, how much performance difference did you observe?

The performance was more or less identical after that change.